### PR TITLE
Implement continuous stream chunking for pre-tokenized C4 MLPerf dataset

### PR DESCRIPTION
**Pull request by snehalv2002**
# Description

### Context & Problem
In the MLCommons DeepSeek-V3 reference specification, pre-training runs on an unpacked, monolithic continuous token stream across documents with 0% padding, monotonic position IDs $0..4095$, unrestricted cross-document causal attention, and 100% loss participation (`eod_mask_loss=False`).

Previously in MaxText, pre-tokenized C4 MLPerf datasets were routed through online sequence bin-packing (`sequence_packing.pack_dataset`). This caused several discrepancies:
1. **Document Isolation:** Documents within a 4096 sequence received discrete segment IDs (`[1, 2, 3, ...]`), blocking attention across document boundaries in `attention_op.py`.
2. **Padding Waste:** Incomplete sequences contained trailing padding slots (`segment_id = 0`), creating ~12.7% padding waste on typical shards.
3. **Position Resets:** Positions reset to 0 at each document boundary instead of indexing monotonically $0..4095$.
4. **Token 0 Deletion Bug in `reduce_concat_tokens`:** The legacy concatenation helper stripped padding via `tf.boolean_mask(tokens, tf.cast(tokens, tf.bool))`. Because casting integer 0 to bool yields `False`, any valid vocabulary token `0` in pre-tokenized data was silently dropped.
5. **Loss Masking of Quotation Marks:** Legacy `format_fn` zeroed segmentation where `targets == eos_id` with default `eos_id=1`. In the LLaMA-3 tokenizer vocabulary, token 1 is `"` (quotation mark), so quotation marks were unintentionally masked from loss computation.

### Solution
- Added `chunk_token_stream()` in `src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py`: Flattens token sequences across document boundaries using `.unbatch().batch(sequence_length, drop_remainder=True)`. This preserves all tokens (including token ID 0) and eliminates padding waste.
- Added `format_continuous_stream_fn()`: Sets monotonic position IDs (`tf.range(max_target_length)`), uniform segment IDs (all `1`s for both inputs and targets), and shifted targets.
- Updated `preprocess_train_dataset()` and `preprocess_eval_dataset()`: Routes `is_tokenized_dataset=True` directly through continuous stream chunking.

# Tests

### 1. Unit & Integration Test Suite
Authored 14 comprehensive tests in `tests/unit/tfds_c4_mlperf_stream_chunking_test.py`:
- `test_chunk_token_stream_preserves_token_zero_and_chunks_perfectly`: Verifies token 0 is preserved and chunks are contiguous.
- `test_chunk_token_stream_drops_remainder_cleanly`: Validates remainder dropping when token counts are not divisible by sequence length.
- `test_chunk_token_stream_handles_empty_documents_gracefully`: Handles 0-length documents without error.
- `test_chunk_token_stream_handles_documents_exceeding_sequence_length`: Tests documents > 4096 tokens (e.g. 10,000 tokens).
- `test_format_continuous_stream_fn_monotonic_positions_and_loss_mask`: Validates monotonic position IDs $0..4095$ and uniform 1s segmentation.
- `test_format_continuous_stream_fn_preserves_llama3_quote_token`: Confirms token ID 1 is not masked from loss.
- `test_preprocess_train_dataset_tokenized_stream_end_to_end` & `test_preprocess_eval_dataset_tokenized_stream_end_to_end`: Tests full data loading pipeline with batching.
- `test_end_to_end_model_forward_and_loss_calculation`: Feeds chunked batches to a MaxText `Transformer` Flax module, verifies 100% loss contribution (`aux["total_weights"] == batch_size * seq_len`), and computes finite gradients via `jax.value_and_grad`.
- `test_real_gcs_dataset_loading`: Direct streaming test loading live shards from `gs://mlperf-6-submission-us-central1/tfds-fixed-reshard/c4/en/3.0.5`.

**Reproduction Command:**
```bash
PYTHONPATH=src python3 -m unittest tests/unit/tfds_c4_mlperf_stream_chunking_test.py -v
```

### 2. End-to-End Execution on Remote CPU VM (`train.py`)
Tested pre-training execution directly with `src/maxtext/trainers/pre_train/train.py` on a remote CPU VM (`anfals-llama405`) pointing at the real C4 MLPerf dataset in GCS (`gs://mlperf-6-submission-us-central1/tfds-fixed-reshard`, `c4/en:3.0.5`):

#### Code Instrumentation
In `src/maxtext/trainers/pre_train/train.py`, right after loading the batch (`data_loader.load_next_batch`):
```python
if step == start_step:
  import os, numpy as np
  mask_path = os.environ.get("DUMP_MASK_PATH", "first_batch_mask.npy")
  np.save(mask_path, np.array(example_batch["inputs_segmentation"]))
  print(f"[MASK_DUMP] Saved first batch mask input to {mask_path} with shape {example_batch['inputs_segmentation'].shape}")
```

#### Execution Command
```bash
./_agents/skills/remote-tpu-run/scripts/remote_run_cpu.sh \
  "DUMP_MASK_PATH=mask.npy ~/maxtext/maxtext_venv/bin/python \
  src/maxtext/trainers/pre_train/train.py \
  src/maxtext/configs/base.yml \
  dataset_type=c4_mlperf \
  dataset_name=c4/en:3.0.5 \
  dataset_path=gs://mlperf-6-submission-us-central1/tfds-fixed-reshard \
  tokenize_train_data=false \
  train_data_columns=\"['ids']\" \
  train_split=train \
  base_num_decoder_layers=2 \
  base_emb_dim=128 \
  base_mlp_dim=256 \
  base_num_query_heads=4 \
  base_num_kv_heads=4 \
  head_dim=32 \
  max_target_length=4096 \
  per_device_batch_size=1 \
  global_batch_size_to_train_on=1 \
  global_batch_size_to_load=1 \
  steps=1 \
  eval_interval=0 \
  enable_checkpointing=false \
  scan_layers=false \
  skip_jax_distributed_system=true"
```

#### Verification Results:
- **`origin/main` Baseline:**
  ```
  completed step: 0, seconds: 4.775, total_weights: 3575, loss: 10.032, lm_loss: 10.032
  [MASK_DUMP] Saved first batch mask input to mask_origin_main.npy with shape (1, 4096)
  ```
  - Contained 8 packed documents with document isolation boundaries.
  - 521 padding tokens (`segment_id = 0`) from index 3575 to 4095 (**12.72% padding waste**).
  - Effective loss tokens: **3,575 / 4,096** (all padding masked).
- **`this branch` (Continuous Stream Chunking):**
  ```
  completed step: 0, seconds: 2.549, total_weights: 4096, loss: 10.098, lm_loss: 10.098
  [MASK_DUMP] Saved first batch mask input to mask_this_branch.npy with shape (1, 4096)
  ```
  - Uniform `segment_id = 1` across all 4096 tokens.
  - 0 padding tokens (**0% padding waste**).
  - Full lower-triangular causal attention across document boundaries.
  - Effective loss tokens: **4,096 / 4,096** (**100% loss participation**, matching MLCommons reference).

[Mask image comparison](https://screenshot-v2.corp.google.com/6j5a2s9qlckd0)

# Checklist

- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py [hunk 1]
```diff
@@ -259,30 +298,44 @@ def preprocess_train_dataset(
...
-  # 3. sequence_packing directly packs pre-tokenized documents preserving
-  #    segment boundaries.
-  if not is_tokenized_dataset:
+    eos_id = 1
+
+  if is_tokenized_dataset:
+    # Continuous stream chunking for pre-tokenized TFDS shards:
+    # 1. Flatten token streams into contiguous max_target_length chunks across document boundaries (0% padding).
+    # 2. Set monotonic position IDs: [0, 1, ..., max_target_length - 1].
+    # 3. Uniform 1s for segmentation to allow cross-document causal attention and 100% loss participation.
```
Comment (dandragona) [inline src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py]: Please update the comment to include the fact that we are adding the eod token.

File: src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py [hunk 2]
```diff
@@ -259,30 +298,44 @@ def preprocess_train_dataset(
...
     pad_id = -1
-
-  # Skip tokenization/chunking for pre-tokenized data (e.g. integer 'ids'):
-  # 1. TokenizeOp expects string inputs, not integer IDs.
-  # 2. reduce_concat_tokens/split_tokens strip padding via bool-cast, dropping
-  #    valid token ID 0.
-  # 3. sequence_packing directly packs pre-tokenized documents preserving
-  #    segment boundaries.
-  if not is_tokenized_dataset:
+    eos_id = 1
```
Comment (dandragona) [inline src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py]: Just curious, how do we know that this is a safe token to use for eos? also can we change it to be eod as I think that is a slightly more specific name to its purpose.

File: src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py [hunk 3]
```diff
@@ -297,8 +350,29 @@ def preprocess_eval_dataset(
     is_tokenized_dataset: bool = True,
 ) -> tf.data.Dataset:
   """Preprocess the evaluation dataset."""
-  # group text up to max_target_length if the dataset is not pre-tokenized/pre-processed
-  if not is_tokenized_dataset:
+  if sp_tokenizer is not None:
```
Comment (dandragona) [inline src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py]: can we DRY this code?

File: tests/unit/tfds_c4_mlperf_stream_chunking_test.py [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+    for chunk in chunks:
+      self.assertEqual(chunk["targets"].shape, (seq_len,))
+
+    reconstructed = np.concatenate([c["targets"] for c in chunks])
+    np.testing.assert_array_equal(reconstructed, np.array(raw_tokens, dtype=np.int32))
+
+    # Explicitly check that token 0 appears in the expected positions
+    self.assertEqual(reconstructed[0], 0)
+    self.assertEqual(reconstructed[3], 0)
+    self.assertEqual(reconstructed[6], 0)
```
Comment (dandragona) [inline tests/unit/tfds_c4_mlperf_stream_chunking_test.py]: My expectation with eod_id = - 1 would be to get exactly 3 sequences:

[0, 10, 20, 0]
[30, -1, 40, 0]
[50, 60, -1, 70]

Is my expectation incorrect?

**Review by dandragona — COMMENTED** [review]
A few comments, mostly just around some confusion I have on end_of_document tokens

File: src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py (hunk 2 above)
Comment (RissyRan) [reply src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py]: Usually should come from tokenzier.

**Review by RissyRan — COMMENTED** [review]
Thanks for the change! Could you also help take a look at Gemini's review and see if tips make sense?

I will point you to the convergence tests.

File: src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py (hunk 1 above)
Comment (snehalv2002) [reply src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py]: done

File: src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py (hunk 2 above)
Comment (snehalv2002) [reply src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py]: changed to get from tokenizer

File: src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py (hunk 3 above)
Comment (snehalv2002) [reply src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py]: decided to remove it entirely since we know the tokenizer will not be none for mlperf runs we don't need all the casework

File: tests/unit/tfds_c4_mlperf_stream_chunking_test.py (hunk 4 above)
Comment (snehalv2002) [reply tests/unit/tfds_c4_mlperf_stream_chunking_test.py]: Just updated the `chunk_token_stream()` function to add `eod_id` after the end of each document. now the test aligns with your expectation

File: src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py (hunk 3 above)
Comment (dandragona) [reply src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py]: Even better :)

**Review by dandragona — APPROVED** [review]

File: src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py [hunk 5]
```diff
@@ -259,30 +307,33 @@ def preprocess_train_dataset(
...
   """Preprocess the training dataset."""
-  if sp_tokenizer.pad_id is not None:
-    pad_id = sp_tokenizer.pad_id
-  elif sp_tokenizer.unk_id is not None:
-    pad_id = sp_tokenizer.unk_id
+  pad_id = sp_tokenizer.pad_id
+  eod_id = sp_tokenizer.eos_id
+
+  if is_tokenized_dataset:
+    # Continuous stream chunking for pre-tokenized TFDS shards:
```
Comment (RissyRan) [inline src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py]: Wondering why the continous stream only for tokenized dataset? Should we consider this as a feature by a flag that support for both tokenized and non-tokenized?

**Review by RissyRan — APPROVED** [review]
Thanks for the change! Maybe let's make unit tests concise? In my opinion, last 2 (test_end_to_end_model_forward_and_loss_calculation, and test_real_gcs_dataset_loading) could be removed.

File: src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py (hunk 5 above)
Comment (snehalv2002) [reply src/maxtext/input_pipeline/tfds_data_processing_c4_mlperf.py]: added a flag to support both tokenized and non-tokenized

**Review by gobbleturk — APPROVED** [review]
